File: app/src/main/java/com/kira/kdownloader/settings/ui/SettingsViewModel.java
package com.kira.kdownloader.settings.ui;

import android.app.Application;
import android.content.Context;
import android.net.Uri;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.kira.kdownloader.BuildConfig;
import com.kira.kdownloader.data.AppDatabase;
import com.kira.kdownloader.settings.AppSettings;
import com.kira.kdownloader.settings.AppearanceSettings;
import com.kira.kdownloader.settings.BehaviorSettings;
import com.kira.kdownloader.settings.DownloadSettings;
import com.kira.kdownloader.settings.HistorySettings;
import com.kira.kdownloader.settings.NetworkSettings;
import com.kira.kdownloader.settings.NotificationSettings;
import com.kira.kdownloader.settings.ProcessingSettings;
import com.kira.kdownloader.settings.SettingsCategory;
import com.kira.kdownloader.settings.SettingsRepository;
import com.kira.kdownloader.settings.StorageSettings;
import com.kira.kdownloader.settings.SubtitleSettings;
import com.kira.kdownloader.settings.platform.FolderAccessManager;
import com.kira.kdownloader.settings.platform.SystemStatus;
import com.kira.kdownloader.settings.store.KeystoreSecureStore;
import com.kira.kdownloader.settings.store.SharedPreferencesKeyValueStore;
import com.kira.kdownloader.util.AppExecutors;

import java.io.File;

public final class SettingsViewModel extends AndroidViewModel {
    private static final String HISTORY_CACHE = "history_cache";
    private static final String KEY_RECENT_URLS = "recent_urls";
    private static final String KEY_SEARCH_HISTORY = "search_history";

    private final SettingsRepository repository;
    private final FolderAccessManager folders;
    private final SystemStatus system;
    private final LiveData<AppSettings> settingsLive;
    private final MutableLiveData<SystemStatus.Snapshot> systemStatus;

    public SettingsViewModel(@NonNull Application application) {
        super(application);
        repository = new SettingsRepository(new SharedPreferencesKeyValueStore(application),
                new KeystoreSecureStore(application));
        folders = new FolderAccessManager(application);
        system = new SystemStatus(application);
        settingsLive = repository.observe();
        systemStatus = new MutableLiveData<>(system.snapshot());
    }

    public LiveData<AppSettings> getSettingsLive() { return settingsLive; }
    public AppSettings getSettingsValue() { return repository.read(); }
    public LiveData<SystemStatus.Snapshot> getSystemStatus() { return systemStatus; }
    public FolderAccessManager getFolders() { return folders; }
    public SystemStatus getSystem() { return system; }
    public void refreshStatus() { systemStatus.setValue(system.snapshot()); }

    public void setDownload(DownloadSettings value) { repository.update(s -> s.withDownload(value)); }
    public void setStorage(StorageSettings value) { repository.update(s -> s.withStorage(value)); }
    public void setBehavior(BehaviorSettings value) { repository.update(s -> s.withBehavior(value)); }
    public void setNetwork(NetworkSettings value) { repository.update(s -> s.withNetwork(value)); }
    public void setSubtitles(SubtitleSettings value) { repository.update(s -> s.withSubtitles(value)); }
    public void setNotifications(NotificationSettings value) { repository.update(s -> s.withNotifications(value)); }
    public void setAppearance(AppearanceSettings value) { repository.update(s -> s.withAppearance(value)); }
    public void setHistory(HistorySettings value) { repository.update(s -> s.withHistory(value)); }
    public void setProcessing(ProcessingSettings value) { repository.update(s -> s.withProcessing(value)); }
    public void setProxyPassword(String value) { repository.setProxyPassword(value); }

    public void onFolderSelected(FolderSlot slot, Uri uri) {
        folders.persist(uri);
        StorageSettings storage = repository.read().getStorage();
        String value = uri.toString();
        if (slot == FolderSlot.DOWNLOAD) storage = storage.withDownloadFolderUri(value);
        else if (slot == FolderSlot.VIDEO) storage = storage.withVideoFolderUri(value);
        else if (slot == FolderSlot.AUDIO) storage = storage.withAudioFolderUri(value);
        else storage = storage.withTempFolderUri(value);
        setStorage(storage);
    }

    public void clearFolder(FolderSlot slot) {
        StorageSettings storage = repository.read().getStorage();
        String current;
        if (slot == FolderSlot.DOWNLOAD) current = storage.getDownloadFolderUri();
        else if (slot == FolderSlot.VIDEO) current = storage.getVideoFolderUri();
        else if (slot == FolderSlot.AUDIO) current = storage.getAudioFolderUri();
        else current = storage.getTempFolderUri();
        folders.release(current);
        if (slot == FolderSlot.DOWNLOAD) storage = storage.withDownloadFolderUri("");
        else if (slot == FolderSlot.VIDEO) storage = storage.withVideoFolderUri("");
        else if (slot == FolderSlot.AUDIO) storage = storage.withAudioFolderUri("");
        else storage = storage.withTempFolderUri("");
        setStorage(storage);
    }

    public void resetCategory(SettingsCategory category) { repository.resetCategory(category); }
    public void resetAll() { repository.resetAll(); }
    public String exportJson() { return repository.exportToJson(); }
    public SettingsRepository.ImportResult importJson(String json) { return repository.importFromJson(json); }

    public long recoverableTempBytes() {
        long total = 0;
        for (File file : cacheDirs()) total += dirSize(file);
        return total;
    }

    public long clearTempFiles() {
        long before = recoverableTempBytes();
        for (File dir : cacheDirs()) {
            File[] files = dir.listFiles();
            if (files != null) for (File file : files) deleteTree(file);
        }
        return before;
    }

    public void clearHistory() {
        AppExecutors.io().execute(() -> AppDatabase.get(getApplication()).downloadDao().clearAll());
    }

    public void clearAllAppData() {
        clearHistory();
        AppExecutors.io().execute(this::clearTempFiles);
        resetAll();
    }

    public void clearRecentUrls() { prefs().edit().remove(KEY_RECENT_URLS).apply(); }
    public void clearSearchHistory() { prefs().edit().remove(KEY_SEARCH_HISTORY).apply(); }

    public String buildDiagnostics() {
        AppSettings value = repository.read();
        return "KDownloader diagnostics\n"
                + "app.version=" + BuildConfig.VERSION_NAME + '\n'
                + "app.build=" + BuildConfig.VERSION_CODE + '\n'
                + "ytdlp.bundled=" + BuildConfig.BUNDLED_YTDLP_VERSION + '\n'
                + "android.sdk=" + android.os.Build.VERSION.SDK_INT + '\n'
                + "device.model=" + android.os.Build.MODEL + '\n'
                + "device.manufacturer=" + android.os.Build.MANUFACTURER + '\n'
                + "log.level=" + value.getProcessing().getLogLevel().getKey() + '\n'
                + "network.allowed=" + value.getNetwork().getAllowedNetworks().getKey() + '\n'
                + "proxy.enabled=" + (!"disabled".equals(value.getNetwork().getProxyType().getKey())) + '\n'
                + "notifications.enabled=" + system.notificationsEnabled() + '\n'
                + "battery.exempt=" + system.ignoringBatteryOptimizations() + '\n'
                + "background.restricted=" + system.backgroundRestricted() + '\n';
    }

    public void clearDiagnosticLog() {
        AppExecutors.io().execute(() -> new File(getApplication().getCacheDir(), "diagnostics.log").delete());
    }

    private android.content.SharedPreferences prefs() {
        return getApplication().getSharedPreferences(HISTORY_CACHE, Context.MODE_PRIVATE);
    }
    private File[] cacheDirs() {
        File external = getApplication().getExternalCacheDir();
        return external == null ? new File[]{getApplication().getCacheDir()}
                : new File[]{getApplication().getCacheDir(), external};
    }
    private static long dirSize(File file) {
        if (file.isFile()) return file.length();
        long total = 0;
        File[] children = file.listFiles();
        if (children != null) for (File child : children) total += dirSize(child);
        return total;
    }
    private static boolean deleteTree(File file) {
        File[] children = file.listFiles();
        if (children != null) for (File child : children) deleteTree(child);
        return !file.exists() || file.delete();
    }
}

File: app/src/main/java/com/kira/kdownloader/settings/ui/SettingsViewModel.kt
package com.kira.kdownloader.settings.ui

import android.app.Application
import android.net.Uri
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.Observer
import androidx.lifecycle.viewModelScope
import com.kira.kdownloader.data.AppDatabase
import com.kira.kdownloader.settings.AppSettings
import com.kira.kdownloader.settings.AppearanceSettings
import com.kira.kdownloader.settings.BehaviorSettings
import com.kira.kdownloader.settings.DownloadSettings
import com.kira.kdownloader.settings.HistorySettings
import com.kira.kdownloader.settings.NetworkSettings
import com.kira.kdownloader.settings.NotificationSettings
import com.kira.kdownloader.settings.ProcessingSettings
import com.kira.kdownloader.settings.SettingsCategory
import com.kira.kdownloader.settings.SettingsRepository
import com.kira.kdownloader.settings.StorageSettings
import com.kira.kdownloader.settings.SubtitleSettings
import com.kira.kdownloader.settings.platform.FolderAccessManager
import com.kira.kdownloader.settings.platform.SystemStatus
import com.kira.kdownloader.settings.store.KeystoreSecureStore
import com.kira.kdownloader.settings.store.SharedPreferencesKeyValueStore
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.io.File

/** Folder slots the user can configure (Section 3). */
enum class FolderSlot { DOWNLOAD, VIDEO, AUDIO, TEMP }

class SettingsViewModel(application: Application) : AndroidViewModel(application) {

    private val repository = SettingsRepository(
        SharedPreferencesKeyValueStore(application),
        KeystoreSecureStore(application),
    )

    val folders = FolderAccessManager(application)
    val system = SystemStatus(application)

    val settingsLive = repository.observe()
    private val settingsState = MutableStateFlow(repository.read())
    private val settingsObserver = Observer<AppSettings> { settingsState.value = it }
    val settings: StateFlow<AppSettings> = settingsState.asStateFlow()

    private val statusState = MutableStateFlow(system.snapshot())
    val systemStatus: StateFlow<SystemStatus.Snapshot> = statusState.asStateFlow()

    init {
        settingsLive.observeForever(settingsObserver)
    }

    override fun onCleared() {
        settingsLive.removeObserver(settingsObserver)
        super.onCleared()
    }

    fun refreshStatus() {
        statusState.value = system.snapshot()
    }

    // ---- Grouped updates (each persists immediately) -------------------------

    fun setDownload(value: DownloadSettings) = repository.update { it.withDownload(value) }
    fun setStorage(value: StorageSettings) = repository.update { it.withStorage(value) }
    fun setBehavior(value: BehaviorSettings) = repository.update { it.withBehavior(value) }
    fun setNetwork(value: NetworkSettings) = repository.update { it.withNetwork(value) }
    fun setSubtitles(value: SubtitleSettings) = repository.update { it.withSubtitles(value) }
    fun setNotifications(value: NotificationSettings) = repository.update { it.withNotifications(value) }
    fun setAppearance(value: AppearanceSettings) = repository.update { it.withAppearance(value) }
    fun setHistory(value: HistorySettings) = repository.update { it.withHistory(value) }
    fun setProcessing(value: ProcessingSettings) = repository.update { it.withProcessing(value) }

    // ---- Proxy credentials ---------------------------------------------------

    fun setProxyPassword(password: String?) = repository.setProxyPassword(password)

    // ---- Folders (SAF) -------------------------------------------------------

    fun onFolderSelected(slot: FolderSlot, uri: Uri) {
        folders.persist(uri)
        val value = uri.toString()
        val storage = settings.value.storage
        setStorage(
            when (slot) {
                FolderSlot.DOWNLOAD -> storage.withDownloadFolderUri(value)
                FolderSlot.VIDEO -> storage.withVideoFolderUri(value)
                FolderSlot.AUDIO -> storage.withAudioFolderUri(value)
                FolderSlot.TEMP -> storage.withTempFolderUri(value)
            },
        )
    }

    fun clearFolder(slot: FolderSlot) {
        val storage = settings.value.storage
        val current = when (slot) {
            FolderSlot.DOWNLOAD -> storage.downloadFolderUri
            FolderSlot.VIDEO -> storage.videoFolderUri
            FolderSlot.AUDIO -> storage.audioFolderUri
            FolderSlot.TEMP -> storage.tempFolderUri
        }
        folders.release(current)
        setStorage(
            when (slot) {
                FolderSlot.DOWNLOAD -> storage.withDownloadFolderUri("")
                FolderSlot.VIDEO -> storage.withVideoFolderUri("")
                FolderSlot.AUDIO -> storage.withAudioFolderUri("")
                FolderSlot.TEMP -> storage.withTempFolderUri("")
            },
        )
    }

    // ---- Reset (Section 13) --------------------------------------------------

    fun resetCategory(category: SettingsCategory) = repository.resetCategory(category)
    fun resetAll() = repository.resetAll()

    // ---- Export / Import (Section 9) -----------------------------------------

    fun exportJson(): String = repository.exportToJson()

    fun importJson(json: String): SettingsRepository.ImportResult = repository.importFromJson(json)

    // ---- Destructive maintenance actions -------------------------------------

    /** Size of removable temporary/cache data (Section 3 "Clear temporary files"). */
    suspend fun recoverableTempBytes(): Long = withContext(Dispatchers.IO) {
        cacheDirs().sumOf { dirSize(it) }
    }

    /** Deletes temporary/cache files. Never touches completed downloads (Section 3). */
    suspend fun clearTempFiles(): Long = withContext(Dispatchers.IO) {
        val before = cacheDirs().sumOf { dirSize(it) }
        cacheDirs().forEach { dir -> dir.listFiles()?.forEach { it.deleteRecursively() } }
        before
    }

    fun clearHistory() {
        viewModelScope.launch(Dispatchers.IO) {
            runCatching { AppDatabase.get(getApplication()).downloadDao().clearAll() }
        }
    }

    /** Resets settings and clears caches + history. Downloaded media files are left in place. */
    fun clearAllAppData() {
        clearHistory()
        viewModelScope.launch(Dispatchers.IO) {
            runCatching { cacheDirs().forEach { d -> d.listFiles()?.forEach { it.deleteRecursively() } } }
        }
        resetAll()
    }

    /** Clears the cached recently-used URLs (Section 9). Does not affect download history. */
    fun clearRecentUrls() {
        getApplication<Application>()
            .getSharedPreferences(HISTORY_CACHE, android.content.Context.MODE_PRIVATE)
            .edit().remove(KEY_RECENT_URLS).apply()
    }

    /** Clears the cached search history (Section 9). */
    fun clearSearchHistory() {
        getApplication<Application>()
            .getSharedPreferences(HISTORY_CACHE, android.content.Context.MODE_PRIVATE)
            .edit().remove(KEY_SEARCH_HISTORY).apply()
    }

    /**
     * Builds a diagnostic report that deliberately excludes credentials, private URLs, folder
     * contents, and PII (Section 12). Only non-identifying app/device metadata is included.
     */
    fun buildDiagnostics(): String {
        val s = settings.value
        return buildString {
            appendLine("KDownloader diagnostics")
            appendLine("app.version=${com.kira.kdownloader.BuildConfig.VERSION_NAME}")
            appendLine("app.build=${com.kira.kdownloader.BuildConfig.VERSION_CODE}")
            appendLine("ytdlp.bundled=${com.kira.kdownloader.BuildConfig.BUNDLED_YTDLP_VERSION}")
            appendLine("android.sdk=${android.os.Build.VERSION.SDK_INT}")
            appendLine("device.model=${android.os.Build.MODEL}")
            appendLine("device.manufacturer=${android.os.Build.MANUFACTURER}")
            appendLine("log.level=${s.processing.logLevel.key}")
            appendLine("network.allowed=${s.network.allowedNetworks.key}")
            appendLine("proxy.enabled=${s.network.proxyType.key != "disabled"}")
            appendLine("notifications.enabled=${system.notificationsEnabled()}")
            appendLine("battery.exempt=${system.ignoringBatteryOptimizations()}")
            appendLine("background.restricted=${system.backgroundRestricted()}")
        }
    }

    fun clearDiagnosticLog() {
        viewModelScope.launch(Dispatchers.IO) {
            runCatching { File(getApplication<Application>().cacheDir, "diagnostics.log").delete() }
        }
    }

    private fun cacheDirs(): List<File> {
        val app = getApplication<Application>()
        return listOfNotNull(app.cacheDir, app.externalCacheDir)
    }

    private fun dirSize(dir: File): Long =
        dir.walkBottomUp().filter { it.isFile }.sumOf { it.length() }

    private companion object {
        const val HISTORY_CACHE = "history_cache"
        const val KEY_RECENT_URLS = "recent_urls"
        const val KEY_SEARCH_HISTORY = "search_history"
    }
}
